File: src/main/java/kr/mjc/rachel/jdbc/user/GetUser.java
package kr.mjc.rachel.jdbc.user;

import kr.mjc.rachel.jdbc.DataSourceFactory;
import kr.mjc.rachel.jdbc.user.User;

import javax.sql.DataSource;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.Scanner;

public class GetUser {
  public static void main(String[] args) {
    System.out.println("Get - id ?");
    Scanner scanner = new Scanner(System.in);
    int id = scanner.nextInt();
    scanner.close();
    final String sql =
        "select id, username, password, first_name, date_joined from user where id=?";
    // dataSource를 구한다.
    DataSource dataSource = DataSourceFactory.getDataSource();
    // try-with-resources 구문. finally 없이 자동 close()
    try (Connection conn = dataSource.getConnection();
        PreparedStatement ps = conn.prepareStatement(sql)) {
      ps.setInt(1, id);
      // preparedStatement를 실행해서 resultSet를 리턴한다.
      ResultSet rs = ps.executeQuery();
      if (rs.next()) {
        // Object - Relational Mapping. 데이터베이스 행으로 오브젝트 만들기
        User user = new User();
        user.setId(rs.getInt("id"));
        user.setUsername(rs.getString("username"));
        user.setFirstName(rs.getString("first_name"));
        user.setDateJoined(rs.getTimestamp("date_joined").toLocalDateTime());
        System.out.println(user);
      } else {
        System.out.println("사용자 없음");
      }
    } catch (SQLException e) {
      throw new RuntimeException(e);
    }
  }
}

File: src/main/kotlin/kr/mjc/jacob/jdbc/user/raw/GetUser.kt
package kr.mjc.jacob.jdbc.user.raw

import kr.mjc.jacob.jdbc.DataSourceFactory
import kr.mjc.jacob.jdbc.user.User
import java.util.*

fun main() {
  print("Get - id ? ")
  val id = Scanner(System.`in`).use { it.nextInt() }

  val sql = "select id, username, first_name, date_joined from user where id=?"

  DataSourceFactory.dataSource.connection.use { conn ->
    conn.prepareStatement(sql).use { ps ->
      ps.setInt(1, id)
      ps.executeQuery().use { rs ->
        if (rs.next()) {
          val user =
            User(id = rs.getInt("id"), username = rs.getString("username"),
                firstName = rs.getString("first_name"),
                dateJoined = rs.getTimestamp("date_joined").toLocalDateTime())
          println(user)
        } else {
          println("사용자 없음")
        }
      }
    }
  }
}
